ui/app/src/components/secrets/SecretDrawer.tsx

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Secret, DispatchWithPromise, Action } from '@perses-dev/core';
+import { Dispatch, DispatchWithoutAction, useState } from 'react';
+import { Drawer, ErrorAlert, ErrorBoundary } from '@perses-dev/components';
+import { DeleteSecretDialog } from '../dialogs';
+import { SecretEditorForm } from './SecretEditorForm';
+
+interface SecretDrawerProps<T extends Secret> {
+  secret: T;
+  isOpen: boolean;
+  action: Action;
+  isReadonly?: boolean;
+  onSave: Dispatch<T>;
+  onDelete?: DispatchWithPromise<T>;
+  onClose: DispatchWithoutAction;
+}
+
+export function SecretDrawer<T extends Secret>(props: SecretDrawerProps<T>) {
+  const { secret, isOpen, action, isReadonly, onSave, onClose, onDelete } = props;
+  const [isDeleteSecretDialogStateOpened, setDeleteSecretDialogStateOpened] = useState<boolean>(false);
+
+  // Disables closing on click out. This is a quick-win solution to avoid losing draft changes.
+  // -> TODO find a way to enable closing by clicking-out in edit view, with a discard confirmation modal popping up
```

Comment:
In future could maybe duplicate / repurpose how discard changes is implemented in panel editor from #834

Reply:
Actually this comment is copy-pasted from VariableDrawer/DatasourceDrawer/etc, we have this problem on all the resources CRUD. Tell me if I'm wrong but from what I can see today we already have the same behavior than for panels here: you have a discard changes dialog showing up when clicking on Cancel, and clicking out of the drawer does nothing. By default clicking out of the drawer closes it & it's quite convenient, so the goal would be to enable this back & thus appropriately open the discard changes dialog when needed. But given how the code is currently structured it's not as simple as it looks like..

Reply:
Ohh my mistake, yeah didn’t realize that was missing, I roughly now how to do it with React Router v5 but let me check with v6 since there were breaking changes